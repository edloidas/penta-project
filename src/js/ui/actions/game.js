// @flow
import { createAction } from 'redux-actions';

// eslint-disable-next-line prettier/prettier
export const resumeGame: PActionCreator<boolean> = createAction('GAME_RESUME');
export const startNewGame: PActionCreator<any> = createAction('GAME_START_NEW');
export const saveGame: PActionCreator<any> = createAction('GAME_SAVE');
export const loadGame: PActionCreator<any> = createAction('GAME_LOAD');
