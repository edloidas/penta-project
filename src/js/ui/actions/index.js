import { createAction } from 'redux-actions';

export const showAbout = createAction('show about');
export const exitGame = createAction('exit game');
