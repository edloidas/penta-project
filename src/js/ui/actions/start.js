// @flow
import { createAction } from 'redux-actions';

// eslint-disable-next-line prettier/prettier
export const finalizeStart: PActionCreator<boolean> = createAction('START_FINALIZE');
export const closeStart: PActionCreator<boolean> = createAction('START_CLOSE');
