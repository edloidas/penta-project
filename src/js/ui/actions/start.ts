import { createAction } from 'redux-actions';

export const finalizeStart: PActionCreator<boolean> = createAction('START_FINALIZE'); // prettier-ignore
export const closeStart: PActionCreator<boolean> = createAction('START_CLOSE');

export type ActionsTypes =
  | ReturnType<typeof finalizeStart>
  | ReturnType<typeof closeStart>;
