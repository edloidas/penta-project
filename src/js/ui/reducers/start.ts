import { handleActions, Action as Action } from 'redux-actions';

export type State = { isReady?: boolean, isClosed?: boolean };
type FinalizeStartAction = Action<boolean>;
type CloseStartAction = Action<boolean>;

type Payload = boolean;

const initialState: State = { isReady: false, isClosed: false };

export default handleActions<State, Payload>(
  {
    START_FINALIZE(state: State, action: FinalizeStartAction): State {
      return Object.assign({}, state, { isReady: action.payload });
    },
    START_CLOSE(state: State, action: CloseStartAction): State {
      return Object.assign({}, state, { isClosed: action.payload });
    }
  },
  initialState
);
