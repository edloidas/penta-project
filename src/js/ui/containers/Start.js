// @flow

import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import Transition from 'react-transition-group/Transition';
import LogoScreen from '../components/screens/LogoScreen';
import * as StartActions from '../actions/startScreen';

class Start extends Component {
  static defaultProps: { isReady: true };

  constructor(props) {
    super(props);
    this.handleAnyPress = this.handleAnyPress.bind(this);
  }

  componentDidMount() {
    // May be replaced by some async works in future
    const minimunAwait = 2000; // ms
    setTimeout(() => this.props.actions.startReady(true), minimunAwait);
  }

  props: {
    isReady: boolean,
    isClosed: boolean,
    actions: {
      startReady: typeof StartActions.startReady,
      closeStart: typeof StartActions.closeStart
    }
  };

  // Absence of `handleAnyPress` type leads to flow errors in constructor.
  handleAnyPress: () => void;

  handleAnyPress(e) {
    e.preventDefault();
    if (this.props.isReady) {
      this.props.actions.closeStart(true);
    }
  }

  render() {
    const { isReady, isClosed } = this.props;
    const hidingDuration = 500;
    return (
      <Transition in={isClosed} timeout={hidingDuration}>
        {state =>
          <div
            tabIndex="0"
            role="button"
            onKeyPress={this.handleAnyPress}
            onClick={this.handleAnyPress}>
            {state === 'entered' ? <Redirect to="/menu" push /> : null}
            <LogoScreen
              isLoading={!isReady}
              isHiding={isClosed}
              duration={hidingDuration}
            />
          </div>}
      </Transition>
    );
  }
}

function mapStateToProps(state) {
  return {
    isReady: state.startScreen.isReady,
    isClosed: state.startScreen.isClosed
  };
}

function mapDispatchToProps(dispatch: Function) {
  return {
    actions: bindActionCreators(StartActions, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Start);
