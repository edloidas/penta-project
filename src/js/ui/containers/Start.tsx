import React from 'react';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import Transition from 'react-transition-group/Transition';
import LogoScreen from '../components/screens/LogoScreen';
import * as StartActions from '../actions/start';
import { State } from '../store';
import Focusable from './Focusable';

type Props = {
  isReady: boolean,
  isClosed: boolean,
  actions: {
    finalizeStart: typeof StartActions.finalizeStart,
    closeStart: typeof StartActions.closeStart
  }
};

class Start extends Focusable<Props> {
  static defaultProps: {
    isReady: true
  };

  constructor(props) {
    super(props);
    this.handleAnyPress = this.handleAnyPress.bind(this);
  }

  componentDidMount() {
    // May be replaced by some async works in future
    const minimunAwait = 2000; // ms
    setTimeout(() => {
      this.props.actions.finalizeStart(true);
      this.focus();
    }, minimunAwait);
  }

  props: Props;

  // Absence of `handleAnyPress` type leads to flow errors in constructor.
  handleAnyPress(e) {
    e.preventDefault();
    if (this.props.isReady) {
      this.props.actions.closeStart(true);
    }
  }

  render() {
    const { isReady, isClosed } = this.props;
    const hidingDuration = 500;
    return (
      <Transition in={isClosed} timeout={hidingDuration}>
        {state => (
          <div
            tabIndex={0}
            role="button"
            ref={div => {
              this.focusable = div;
            }}
            onKeyPress={this.handleAnyPress}
            onClick={this.handleAnyPress}>
            {state === 'entered' ? <Redirect to="/menu" push /> : null}
            <LogoScreen
              isLoading={!isReady}
              isHiding={isClosed}
              duration={hidingDuration}
            />
          </div>
        )}
      </Transition>
    );
  }
}

function mapStateToProps(state: State) {
  return {
    isReady: state.start.isReady,
    isClosed: state.start.isClosed
  };
}

function mapDispatchToProps(dispatch: Dispatch<any>) {
  return {
    actions: bindActionCreators(StartActions, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Start);
