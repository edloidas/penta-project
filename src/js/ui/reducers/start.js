// @flow

import { handleActions, type ActionType } from 'redux-actions';
import { finalizeStart, closeStart } from '../actions/start';

export type State = { isReady?: boolean, isClosed?: boolean };
type FinalizeStartAction = ActionType<typeof finalizeStart>;
type CloseStartAction = ActionType<typeof closeStart>;

const initialState: State = { isReady: false, isClosed: false };

export default handleActions(
  {
    START_FINALIZE(state: State, action: FinalizeStartAction): State {
      return Object.assign({}, state, { isReady: action.payload });
    },
    START_CLOSE(state: State, action: CloseStartAction): State {
      return Object.assign({}, state, { isClosed: action.payload });
    }
  },
  initialState
);
