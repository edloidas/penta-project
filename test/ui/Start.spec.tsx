import * as React from 'react';
import { mount } from '../wrap';
import configureStore from '../../src/js/ui/store';
import Start from '../../src/js/ui/containers/Start';
import { finalizeStart } from '../../src/js/ui/actions/start';

describe('<Start />', () => {
  test('Should renders', () => {
    const tree = mount(<Start />);
    expect(tree).toMatchSnapshot();
  });

  test('Should switch to Await loader', () => {
    const store = configureStore();
    store.dispatch(finalizeStart(true));

    const tree = mount(<Start />, store);
    expect(tree).toMatchSnapshot();
  });
});
