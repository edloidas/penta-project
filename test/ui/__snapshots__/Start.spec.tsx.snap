// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Start /> Should renders 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <ConnectedRouter
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(Start)>
        <Start
          actions={
            Object {
              "closeStart": [Function],
              "finalizeStart": [Function],
            }
          }
          isClosed={false}
          isReady={false}
        >
          <Transition
            appear={false}
            enter={true}
            exit={true}
            in={false}
            mountOnEnter={false}
            onEnter={[Function]}
            onEntered={[Function]}
            onEntering={[Function]}
            onExit={[Function]}
            onExited={[Function]}
            onExiting={[Function]}
            timeout={500}
            unmountOnExit={false}
          >
            <div
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex={0}
            >
              <LogoScreen
                duration={500}
                isHiding={false}
                isLoading={true}
              >
                <Styled(styled.div)
                  className="font__mono"
                  theme={
                    Object {
                      "duration": 500,
                      "isHiding": false,
                    }
                  }
                >
                  <styled.div
                    className="font__mono sc-dnqmqq jtumug"
                    theme={
                      Object {
                        "duration": 500,
                        "isHiding": false,
                      }
                    }
                  >
                    <div
                      className="font__mono sc-dnqmqq jtumug sc-bdVaJa iVLFuM"
                    >
                      <Logo
                        renderAuthor={true}
                        theme={
                          Object {
                            "fontSize": 1,
                          }
                        }
                      >
                        <styled.div
                          className="font__mono"
                          theme={
                            Object {
                              "fontSize": 1,
                            }
                          }
                        >
                          <div
                            className="font__mono sc-EHOje kLsBzw"
                          >
                            <styled.div>
                              <div
                                className="sc-bZQynM ixtovb"
                              >
                                edloidas production
                              </div>
                            </styled.div>
                            <styled.div>
                              <div
                                className="sc-gzVnrw MmFrt"
                              >
                                Penta Project
                              </div>
                            </styled.div>
                            <styled.div>
                              <div
                                className="sc-htoDjs ddYdqL"
                              >
                                A Cyberpunk Action Game
                              </div>
                            </styled.div>
                          </div>
                        </styled.div>
                      </Logo>
                      <styled.div>
                        <div
                          className="sc-iwsKbI jRAMkw"
                        >
                          <InfiniteLoader
                            theme={
                              Object {
                                "bg": "#f5f5f5",
                              }
                            }
                          >
                            <styled.ul>
                              <ul
                                className="sc-bxivhb mKstk"
                              >
                                <styled.li
                                  theme={
                                    Object {
                                      "bg": "#f5f5f5",
                                      "delay": "0.2s",
                                    }
                                  }
                                >
                                  <li
                                    className="sc-ifAKCX iUWNCO"
                                  />
                                </styled.li>
                                <styled.li
                                  theme={
                                    Object {
                                      "bg": "#f5f5f5",
                                      "delay": "0.4s",
                                    }
                                  }
                                >
                                  <li
                                    className="sc-ifAKCX gadhOi"
                                  />
                                </styled.li>
                                <styled.li
                                  theme={
                                    Object {
                                      "bg": "#f5f5f5",
                                      "delay": "0.6s",
                                    }
                                  }
                                >
                                  <li
                                    className="sc-ifAKCX jEAgHN"
                                  />
                                </styled.li>
                                <styled.li
                                  theme={
                                    Object {
                                      "bg": "#f5f5f5",
                                      "delay": "0.8s",
                                    }
                                  }
                                >
                                  <li
                                    className="sc-ifAKCX eFvPdM"
                                  />
                                </styled.li>
                              </ul>
                            </styled.ul>
                          </InfiniteLoader>
                        </div>
                      </styled.div>
                    </div>
                  </styled.div>
                </Styled(styled.div)>
              </LogoScreen>
            </div>
          </Transition>
        </Start>
      </Connect(Start)>
    </Router>
  </ConnectedRouter>
</Provider>
`;

exports[`<Start /> Should switch to Await loader 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <ConnectedRouter
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(Start)>
        <Start
          actions={
            Object {
              "closeStart": [Function],
              "finalizeStart": [Function],
            }
          }
          isClosed={false}
          isReady={true}
        >
          <Transition
            appear={false}
            enter={true}
            exit={true}
            in={false}
            mountOnEnter={false}
            onEnter={[Function]}
            onEntered={[Function]}
            onEntering={[Function]}
            onExit={[Function]}
            onExited={[Function]}
            onExiting={[Function]}
            timeout={500}
            unmountOnExit={false}
          >
            <div
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex={0}
            >
              <LogoScreen
                duration={500}
                isHiding={false}
                isLoading={false}
              >
                <Styled(styled.div)
                  className="font__mono"
                  theme={
                    Object {
                      "duration": 500,
                      "isHiding": false,
                    }
                  }
                >
                  <styled.div
                    className="font__mono sc-dnqmqq jtumug"
                    theme={
                      Object {
                        "duration": 500,
                        "isHiding": false,
                      }
                    }
                  >
                    <div
                      className="font__mono sc-dnqmqq jtumug sc-bdVaJa iVLFuM"
                    >
                      <Logo
                        renderAuthor={true}
                        theme={
                          Object {
                            "fontSize": 1,
                          }
                        }
                      >
                        <styled.div
                          className="font__mono"
                          theme={
                            Object {
                              "fontSize": 1,
                            }
                          }
                        >
                          <div
                            className="font__mono sc-EHOje kLsBzw"
                          >
                            <styled.div>
                              <div
                                className="sc-bZQynM ixtovb"
                              >
                                edloidas production
                              </div>
                            </styled.div>
                            <styled.div>
                              <div
                                className="sc-gzVnrw MmFrt"
                              >
                                Penta Project
                              </div>
                            </styled.div>
                            <styled.div>
                              <div
                                className="sc-htoDjs ddYdqL"
                              >
                                A Cyberpunk Action Game
                              </div>
                            </styled.div>
                          </div>
                        </styled.div>
                      </Logo>
                      <styled.div>
                        <div
                          className="sc-iwsKbI jRAMkw"
                        >
                          <AwaitInputLoader
                            theme={
                              Object {
                                "color": "#c5c8c6",
                              }
                            }
                          >
                            <styled.button
                              autoFocus={true}
                              className="font__mono"
                              theme={
                                Object {
                                  "color": "#c5c8c6",
                                }
                              }
                            >
                              <button
                                autoFocus={true}
                                className="font__mono sc-htpNat glyIiT"
                              >
                                Press Any Key
                              </button>
                            </styled.button>
                          </AwaitInputLoader>
                        </div>
                      </styled.div>
                    </div>
                  </styled.div>
                </Styled(styled.div)>
              </LogoScreen>
            </div>
          </Transition>
        </Start>
      </Connect(Start)>
    </Router>
  </ConnectedRouter>
</Provider>
`;
